x/evm/keeper/querier.go

```diff
@@ -88,6 +95,131 @@ func NewQuerier(k types.BaseKeeper, s types.Signer, n types.Nexus) sdk.Querier {
 	}
 }
 
+// QueryPendingCommands returns the pending commands for a gateway
+func QueryPendingCommands(ctx sdk.Context, keeper types.ChainKeeper, n types.Nexus) ([]byte, error) {
+	var resp types.QueryPendingCommandsResponse
+	commands := keeper.GetPendingCommands(ctx)
+
+	for _, cmd := range commands {
+		cmdResp, err := GetCommandResponse(ctx, keeper.GetName(), n, cmd)
+		if err != nil {
+			return nil, sdkerrors.Wrap(types.ErrEVM, err.Error())
+		}
+
+		resp.Commands = append(resp.Commands, cmdResp)
+	}
+
+	return resp.Marshal()
+}
+
+func queryCommand(ctx sdk.Context, keeper types.ChainKeeper, n types.Nexus, id string) ([]byte, error) {
+	cmdID, err := types.HexToCommandID(id)
+	if err != nil {
+		return nil, sdkerrors.Wrap(types.ErrEVM, err.Error())
+	}
+
+	cmd, ok := keeper.GetCommand(ctx, cmdID)
+	if !ok {
+		return nil, sdkerrors.Wrap(types.ErrEVM, fmt.Sprintf("could not find command '%s'", cmd.ID.Hex()))
+	}
+
+	resp, err := GetCommandResponse(ctx, keeper.GetName(), n, cmd)
+	if err != nil {
+		return nil, sdkerrors.Wrap(types.ErrEVM, err.Error())
+	}
+
+	return resp.Marshal()
+}
+
+// GetCommandResponse converts a Command into a CommandResponse type
+func GetCommandResponse(ctx sdk.Context, chainName string, n types.Nexus, cmd types.Command) (types.QueryCommandResponse, error) {
```

Comment:
can't we add something like `AsMap()` to each of these commands instead of having to externally parse them?

Reply:
I thought about doing something like that, but I thought that would be a tacit condoning of using such function anywhere -- and since maps are a non-deterministic data struct, I figured it would be safer to implement them on the querier, since its the one place that non-determinism won't be a major issue.

Reply:
since it's only used in the query, I wouldn't spend more time on this for the time being. Just keep in mind for the future that an interface function implementation is more robust than a switch on the specific type